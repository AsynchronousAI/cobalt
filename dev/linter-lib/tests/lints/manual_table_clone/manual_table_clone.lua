local new1 = {}
for key, value in pairs(stuff) do
	new1[key] = value -- fail
end

local new2 = {}
for key, value in ipairs(stuff) do
	new2[key] = value -- fail
end

local new3 = {}
for key, value in next, stuff do
	new3[key] = value -- fail
end

local new4 = {}
for key, value in stuff do
	new4[key] = value -- fail
end

local new5 = {}
for key, value in pairs(stuff) do
	if key == "foo" then
		new5[key] = value -- pass
	end
end

local new6 = {}
new6.used = "welp"
for key, value in pairs(stuff) do
	new6[key] = value -- pass
end

local new7 = {}
for key, value in pairs(stuff) do
	new7[key] = value -- fail
end
new7.used = "too late"

local new8 = {}
for key, value in pairs(getStuff()) do
	new8[key] = value -- fail
end

local new9 = {}

-- For Luau, this can be treated the same as looping over a table result
for key, value in what(stuff) do
	new9[key] = value -- fail
end

local new10 = {}
for key, value in pairs(stuff), what(stuff) do
	new10[key] = value -- shrug
end

local new11 = {}
for key, value in what(stuff), pairs(stuff) do
	new11[key] = value -- shrug
end

local new12 = {}
for key, value in pairs(what)(the) do
	new12[key] = value -- shrug
end

for key, value in pairs(stuff) do
	no()[key] = value -- pass
end

for key, value in pairs(stuff) do
	too.bad[key] = value -- pass
end

for key, value in pairs(stuff) do
	global[key] = value -- pass
end

local new13, new14 = {}, {}
for key, value in pairs(stuff) do
	new13[key], new14[key] = value, -value -- pass
end

local new15 = { x = 1 }
for key, value in pairs(stuff) do
	new15[key] = value -- pass
end

local new16 = {}
whoKnows(new16)
for key, value in pairs(stuff) do
	new16[key] = value -- pass
end

local new17 = whoKnows()
for key, value in pairs(stuff) do
	new17[key] = value -- pass
end

local new18 = {}
for key, value in what, stuff do
	new18[key] = value -- pass
end

local new19 = {}
blaBlaBla()
someStuffHere()
for key, value in pairs(stuff) do
	new19[key] = value -- fail
end

-- weird but valid
local function ipairs(_) end
local newWeirdIpairs = {}
for key, value in ipairs(stuff) do
	newWeirdIpairs[key] = value -- fail, but don't report as ipairs
end
