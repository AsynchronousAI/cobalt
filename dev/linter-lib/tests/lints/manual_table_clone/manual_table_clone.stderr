error[manual_table_clone]: manual implementation of table.clone
  ┌─ manual_table_clone.lua:1:1
  │  
1 │ ╭ local new1 = {}
2 │ │ for key, value in pairs(stuff) do
3 │ │     new1[key] = value -- fail
4 │ │ end
  │ ╰───^
  │  
  = try `local new1 = table.clone(stuff)`

error[manual_table_clone]: manual implementation of table.clone
  ┌─ manual_table_clone.lua:6:1
  │  
6 │ ╭ local new2 = {}
7 │ │ for key, value in ipairs(stuff) do
8 │ │     new2[key] = value -- fail
9 │ │ end
  │ ╰───^
  │  
  = try `local new2 = table.clone(stuff)`
  = if this is a mixed table, then table.clone is not equivalent, as ipairs only goes over the array portion.
    ignore this lint with `-- selene: allow(manual_table_clone)` if this is the case.

error[manual_table_clone]: manual implementation of table.clone
   ┌─ manual_table_clone.lua:11:1
   │  
11 │ ╭ local new3 = {}
12 │ │ for key, value in next, stuff do
13 │ │     new3[key] = value -- fail
14 │ │ end
   │ ╰───^
   │  
   = try `local new3 = table.clone(stuff)`

error[manual_table_clone]: manual implementation of table.clone
   ┌─ manual_table_clone.lua:16:1
   │  
16 │ ╭ local new4 = {}
17 │ │ for key, value in stuff do
18 │ │     new4[key] = value -- fail
19 │ │ end
   │ ╰───^
   │  
   = try `local new4 = table.clone(stuff)`

error[manual_table_clone]: manual implementation of table.clone
   ┌─ manual_table_clone.lua:34:1
   │  
34 │ ╭ local new7 = {}
35 │ │ for key, value in pairs(stuff) do
36 │ │     new7[key] = value -- fail
37 │ │ end
   │ ╰───^
   │  
   = try `local new7 = table.clone(stuff)`

error[manual_table_clone]: manual implementation of table.clone
   ┌─ manual_table_clone.lua:40:1
   │  
40 │ ╭ local new8 = {}
41 │ │ for key, value in pairs(getStuff()) do
42 │ │     new8[key] = value -- fail
43 │ │ end
   │ ╰───^
   │  
   = try `local new8 = table.clone(getStuff())`

error[manual_table_clone]: manual implementation of table.clone
   ┌─ manual_table_clone.lua:45:1
   │  
45 │ ╭ local new9 = {}
46 │ │ 
47 │ │ -- For Luau, this can be treated the same as looping over a table result
48 │ │ for key, value in what(stuff) do
49 │ │     new9[key] = value -- fail
50 │ │ end
   │ ╰───^
   │  
   = try `local new9 = table.clone(what(stuff))`

error[manual_table_clone]: manual implementation of table.clone
    ┌─ manual_table_clone.lua:108:1
    │  
105 │   local new19 = {}
    │   ---------------- remove this definition
    ·  
108 │ ╭ for key, value in pairs(stuff) do
109 │ │     new19[key] = value -- fail
110 │ │ end
    │ ╰───^
    │  
    = try `local new19 = table.clone(stuff)`

error[manual_table_clone]: manual implementation of table.clone
    ┌─ manual_table_clone.lua:114:1
    │  
114 │ ╭ local newWeirdIpairs = {}
115 │ │ for key, value in ipairs(stuff) do
116 │ │     newWeirdIpairs[key] = value -- fail, but don't report as ipairs
117 │ │ end
    │ ╰───^
    │  
    = try `local newWeirdIpairs = table.clone(stuff)`
    = if this is a mixed table, then table.clone is not equivalent, as ipairs only goes over the array portion.
      ignore this lint with `-- selene: allow(manual_table_clone)` if this is the case.

